README.md
# microsim (Microservice Simulator)

The purpose of this project is to be able to easily set up kubernetes deployments with somewhat realistic/controllable traffic and attacks to test load balancing, WAF, and other security solutions in the cluster.

Also available on Docker Hub:

https://hub.docker.com/r/kellybrazil/microsimserver

https://hub.docker.com/r/kellybrazil/microsimclient

This application generates traffic between simulated microservices using JSON over HTTP. Many parameters can be set on the client and server apps (`microsimclient` and `microsimserver`) to specify the listening port, the amount of traffic generated, whether requests are sent to the internet, whether certain application attacks are sent, and whether malware is downloaded. You can also set the client and server to kill themselves after a number of seconds.

Sample Kubernetes deployments can be found in the https://github.com/kellyjonbrazil/microsim/tree/master/k8s_deployments folder.

![Simple Deployment](https://github.com/kellyjonbrazil/microsim/blob/master/k8s_deployments/images/simple.png)

## Usage

All parameters are set via environment variables. Unset parameters will use the defaults:

### microsimserver:

| Parameter          | Values          | Default          | Description                                     |
| ------------------ |-----------------| -----------------| ------------------------------------------------|
| `LISTEN_PORT`      | `1` - `65535`   | `8080`           | Port the HTTP server listens on                 |
| `STATS_PORT`       | `1` - `65535`   | None             | Enable the HTTP stats server running on the specified port |
| `STATSD_HOST`      | `"1.2.3.4"`     | None             | Enable sending StatsD stats to the specified host |
| `STATSD_PORT`      | `1` - `65535`   | `8125`           | Modify the default StatsD destination port, if desired |
| `RESPOND_BYTES`    | `1` - ?         | `16384`          | How many data bytes are added to the response   |
| `STOP_SECONDS`     | `0` - ?         | `0` (never stop) | Kill the server after x seconds                 |
| `STOP_PADDING`     | `True`/`False`  | `False`          | Add random padding to `STOP_SECONDS`            |

### microsimclient:

| Parameter             | Values          | Default  | Description                                     |
| --------------------- |-----------------| ---------| ------------------------------------------------|
| `STATS_PORT`          | `1` - `65535`   | None     | Enable the HTTP stats server running on the specified port |
| `STATSD_HOST`         | `"1.2.3.4"`     | None             | Enable sending StatsD stats to the specified host |
| `STATSD_PORT`         | `1` - `65535`   | `8125`           | Modify the default StatsD destination port, if desired |
| `REQUEST_URLS`          | `"http://auth:8080,http://db:8080"` | None      | One or more comma separated URLs to send requests to. *Note: this is a required parameter* |
| `REQUEST_INTERNET`      | `True`/`False`      | `False`    | Send regular requests to the internet if True             |
| `REQUEST_MALWARE`       | `True`/`False`      | `False`    | Occasionally download an eicar sample from the internet   |
| `SEND_SQLI`             | `True`/`False`      | `False`    | Occasionally send SQLi to the REQUEST_URLS                |
| `SEND_XSS`              | `True`/`False`      | `False`    | Occasionally send XSS to the REQUEST_URLS                 |
| `SEND_DIR_TRAVERSAL`    | `True`/`False`      | `False`    | Occasionally send Directory Traversal to the REQUEST_URLS |
| `SEND_DGA`              | `True`/`False`      | `False`    | Occasionally send DGA DNS requests to the resolver        |
| `REQUEST_WAIT_SECONDS`  | `0` - ? (float)  | `1.0`        | Number of seconds to wait between request loop runs |
| `REQUEST_BYTES`         | `1` - `7980`        | `1024`     | How many data bytes are added to the request              |
| `REQUEST_PROBABILITY`   | `0.9` (float)       | `1.0`      | Float value representing the percentage probability of an internal request triggering per loop |
| `EGRESS_PROBABILITY`    | `0.1` (float)       | `0.1`      | Float value representing the percentage probability of an egress Internet request triggering per loop |
| `ATTACK_PROBABILITY`    | `0.01` (float)      | `0.01`     | Float value representing the percentage probability of one of the attack behaviors triggering per loop |
| `STOP_SECONDS`          | `0` - ?             | `0` (never stop) | Kill the client after x seconds     |
| `STOP_PADDING`          | `True`/`False`      | `False`    | Add random padding to `STOP_SECONDS`            |

### Example docker commands:
```
docker run -d --rm \
    -e LISTEN_PORT=8080 \
    -e STATS_PORT=5001 \
    -e STATSD_HOST="172.17.0.2" \
    -e STATSD_PORT=8100 \
    -e RESPOND_BYTES=1024 \
    -e STOP_SECONDS=300 \
    -e STOP_PADDING=True \
    -p 8080:8080 \
    -p 5001:5001 \
    kellybrazil/microsimserver
```

```
docker run -d --rm \
    -e STATS_PORT=5000 \
    -e STATSD_HOST="172.17.0.2" \
    -e STATSD_PORT=8100 \
    -e REQUEST_URLS="http://service1.default.svc.cluster.local:8080,http://service2.default.svc.cluster.local:5000" \
    -e REQUEST_INTERNET=True \
    -e REQUEST_MALWARE=True \
    -e SEND_SQLI=True \
    -e SEND_DIR_TRAVERSAL=True \
    -e SEND_XSS=True \
    -e SEND_DGA=True \
    -e REQUEST_WAIT_SECONDS=0.5 \
    -e REQUEST_BYTES=128 \
    -e REQUEST_PROBABILITY=0.9 \
    -e EGRESS_PROBABILITY=0.3 \
    -e ATTACK_PROBABILITY=0.05 \
    -e STOP_SECONDS=300 \
    -e STOP_PADDING=True \
    -p 5000:5000 \
    kellybrazil/microsimclient
```

## Example response from the microsimserver:
```
$ curl -X POST localhost:8080/any_arbitrary_path | jq .
{
  "data": "3BbTvDWTD7yWBsWkffiTft5V875EpldKSehv2uXvUSHgl6Sjjl",
  "time": "Sat Aug 17 15:34:09 2019",
  "hostname": "laptop.local",
  "ip": "192.168.1.221",
  "remote": "127.0.0.1",
  "hostheader": "localhost:8080",
  "path": "/any_arbitrary_path"
}
```
The `data` field is randomly generated text. The number of characters is controlled with the `RESPOND_BYTES` parameter.

If a normal `GET` request is sent, then the response will look like this:
```
$ curl localhost:8080/any_arbitrary_path
f0QRurmXLiuungowUkQMJf3z687BDgsUpV3Lupaln2WUH4egk007En3m11lIurSwcekiI1PqhyRHpzPzYB
Sat Aug 17 15:42:29 2019   hostname: laptop.local   ip: 192.168.1.221   remote: 127.0.0.1   hostheader: localhost:8080   path: /any_arbitrary_path
```

In this case, the first line is the `data` field sent from the `microsimserver`.

The `microsimclient` similarly POST's a `data` field in the request body which is filled with randomly generated text. The number of characters is controled with the `REQUEST_BYTES` parameter.

## Local Statistics

If desired, a realtime statistics HTTP server can be enabled by specifying the `STATS_PORT` parameter for both the client and the server. The curl command is included in the Docker container so local and remote stats can also be pulled from within the container. Example output:

### microsimserver:
```
$ curl localhost:5001
{
  "time": "Mon Aug 26 15:26:18 2019",
  "runtime": 42,
  "hostname": "server01",
  "ip": "192.168.1.221",
  "stats": {
    "Requests": 138,
    "Sent Bytes": 2284728,
    "Received Bytes": 142968,
    "Attacks": 2,
    "SQLi": 1,
    "XSS": 0,
    "Directory Traversal": 1
  },
  "config": {
    "LISTEN_PORT": 8080,
    "STATS_PORT": 5001,
    "STATSD_HOST": null,
    "STATSD_PORT": 8125,
    "RESPOND_BYTES": 16384,
    "STOP_SECONDS": 60,
    "STOP_PADDING": true,
    "TOTAL_STOP_SECONDS": 102
  }
}
```

### microsimclient:
```
$ curl localhost:5000
{
  "time": "Mon Aug 26 15:10:40 2019",
  "runtime": 40,
  "hostname": "client01",
  "ip": "192.168.1.221",
  "stats": {
    "Requests": 125,
    "Sent Bytes": 129500,
    "Received Bytes": 2216193,
    "Internet Requests": 12,
    "Attacks": 0,
    "SQLi": 0,
    "XSS": 0,
    "Directory Traversal": 0,
    "DGA": 0,
    "Malware": 0,
    "Error": 3
  },
  "config": {
    "STATS_PORT": 5000,
    "STATSD_HOST": null,
    "STATSD_PORT": 8125,
    "REQUEST_URLS": "http://127.0.0.1:8080",
    "REQUEST_INTERNET": true,
    "REQUEST_MALWARE": true,
    "SEND_SQLI": true,
    "SEND_DIR_TRAVERSAL": true,
    "SEND_XSS": true,
    "SEND_DGA": true,
    "REQUEST_WAIT_SECONDS": 0.3,
    "REQUEST_BYTES": 1024,
    "STOP_SECONDS": 60,
    "STOP_PADDING": true,
    "TOTAL_STOP_SECONDS": 75,
    "REQUEST_PROBABILITY": 1.0,
    "EGRESS_PROBABILITY": 0.1,
    "ATTACK_PROBABILITY": 0.01
  }
}
```

## StatsD Statistics

By configuring the `STATSD_HOST` and the optional `STATSD_PORT` (default UDP port is `8125`) parameter the client and server will send regular request, response, attack, and error stats to the StatsD server for aggregation and graphing.

## Logging

Both the client and server log output to stdout, which will show up in docker and kubernetes logs. The logs will summarize the requests and errors and will also print a summary every 30 seconds of total stats and the stats for the last 30 seconds.

### microsimclient:
```
2019-09-30T16:53:50   Request to http://localhost:8080/   Request size: 1036   Response size: 100172
2019-09-30T16:53:50   Directory Traversal sent: http://localhost:8080/?username=joe%40example.com&password=..%2F..%2F..%2F..%2F..%2Fpasswd
2019-09-30T16:53:50   Internet request to: http://mirror.grid.uchicago.edu/pub/linux/centos/
2019-09-30T16:53:51   Request to http://localhost:8080/   Request size: 1036   Response size: 100172
{"Total": {"Requests": 545, "Sent Bytes": 564620, "Received Bytes": 55301055, "Internet Requests": 58, "Attacks": 5, "SQLi": 0, "XSS": 0, "Directory Traversal": 2, "DGA": 3, "Malware": 0, "Error": 0}, "Last 30 Seconds": {"Requests": 42, "Sent Bytes": 43512, "Received Bytes": 4269205, "Internet Requests": 6, "Attacks": 1, "SQLi": 0, "XSS": 0, "Directory Traversal": 0, "DGA": 1, "Malware": 0, "Error": 0}}
2019-09-30T16:53:51   Request to http://localhost:8080/   Request size: 1036   Response size: 100172
2019-09-30T16:53:51   Request to http://localhost:8080/   Request size: 1036   Response size: 100172
2019-09-30T16:53:52   Request to http://localhost:8080/   Request size: 1036   Response size: 100172
2019-09-30T16:53:52   DGA query sent: xdcc5481252db5f38d5fc18c9ad3b2f7fd.cn   Response: 23.202.231.169
2019-09-30T16:53:52   Request to http://localhost:8080/   Request size: 1036   Response size: 100172
2019-09-30T16:53:53   Request to http://localhost:8080/   Request size: 1036   Response size: 100172
2019-09-30T16:53:53   Request to http://localhost:8080/   Request size: 1036   Response size: 100172
2019-09-30T16:53:53   Request to http://localhost:8080/   Request size: 1036   Response size: 100172
2019-09-30T16:53:54   Internet request to: http://mirrors.oit.uci.edu/centos/
2019-09-30T16:53:54   Request to http://localhost:8080/   Request size: 1036   Response size: 100172
2019-09-30T16:53:54   Request to http://localhost:8080/   Request size: 1036   Response size: 100172
```

### microsimserver:
```
127.0.0.1 - - [26/Aug/2019 15:30:01] "POST / HTTP/1.1" 200 -
127.0.0.1 - - [26/Aug/2019 15:30:02] "POST / HTTP/1.1" 200 -
127.0.0.1 - - [26/Aug/2019 15:30:02] "GET /?username=joe%40example.com&password=pwd%3Cscript%3Ealert%28%27attacked%27%29%3C%2Fscript%3E HTTP/1.1" 200 -
2019-09-30T16:53:50   XSS attack detected
127.0.0.1 - - [26/Aug/2019 15:30:02] "POST / HTTP/1.1" 200 -
127.0.0.1 - - [26/Aug/2019 15:30:02] "POST / HTTP/1.1" 200 -
{"Total": {"Requests": 391, "Sent Bytes": 6473451, "Received Bytes": 404040, "Attacks": 1, "SQLi": 0, "XSS": 1, "Directory Traversal": 0}, "Last 30 Seconds": {"Requests": 97, "Sent Bytes": 1605932, "Received Bytes": 100492, "Attacks": 0, "SQLi": 0, "XSS": 0, "Directory Traversal": 0}}
127.0.0.1 - - [26/Aug/2019 15:30:03] "POST / HTTP/1.1" 200 -
127.0.0.1 - - [26/Aug/2019 15:30:03] "POST / HTTP/1.1" 200 -
```


microsimclient/changelog.txt
microsimclient changelog

20191001 v0.6
- Add thread locking
- Add datetime to log messages
- Add request timeouts

20190827 v0.5
- Fix probabilities for attacks and internal requests
- Add REQUEST_PROBABILITY parameter to control internal requests
- Add more configuration information to HTTP stats
- Make all requests threaded to prevent blocking
- Add STOP_PADDING parameter to add random seconds to STOP_SECONDS

20190821 v0.3
- Change REQUEST_WAIT_SECONDS to float
- Fix parameter boolean logic
- Add stats http server
- Add statsd client
- Add stats to log
- Change ATTACK_PROBABILITY to 0.01
- Change DGA to random choice of one request
- Change egress internet to random choice
- Add curl to container

20190818 v0.2
- Add ATTACK_PROBABILITY parameter
- Add EGRESS_PROBABILITY parameter

20190817 v0.1
- Initial release

microsimserver/microsimservermac.py
#!/usr/local/bin/python3
import sys
import os
import socket
import time
import random
import string
import json
from http.server import BaseHTTPRequestHandler, HTTPServer

LISTEN_PORT = int(os.getenv('LISTEN_PORT', 8080))
RESPOND_BYTES = int(os.getenv('RESPOND_BYTES', 16384))
STOP_SECONDS = int(os.getenv('STOP_SECONDS', 0))

START_TIME = int(time.time())
HOST_NAME = ''

def keep_running():
    if (STOP_SECONDS != 0) and ((START_TIME + STOP_SECONDS) < int(time.time())):
        sys.exit('Server killed after ' + str(STOP_SECONDS) + ' seconds.')
    return True

def insert_data():
    return ''.join([random.choice(string.ascii_letters + string.digits) for n in range(RESPOND_BYTES)])

class httpd(BaseHTTPRequestHandler):
    server_name = socket.gethostname()
    # for mac compatibility
    server_name = server_name + '.local'
    server_ip = socket.gethostbyname(server_name)

    def do_GET(self):
        """simple http response"""
        host_header = self.headers['Host']
        self.send_response(200)
        self.send_header('Content-Type', 'text/html')
        self.end_headers()
        data = insert_data() + '\n'
        info = time.asctime() + '   hostname: ' + self.server_name + '   ip: ' + self.server_ip + '   remote: ' + self.address_string() + '   hostheader: ' + str(host_header) + '   path: ' + self.path + '\n'
        body = data + info
        self.wfile.write(body.encode('utf-8'))

    def do_POST(self):
        """json api response"""
        host_header = self.headers['Host']
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        self.response = {
            'time': time.asctime(),
            'hostname': self.server_name,
            'ip': self.server_ip,
            'remote': self.address_string(),
            'hostheader': host_header,
            'path': self.path,
            'data': insert_data()
        }
        body = json.dumps(self.response)
        self.wfile.write(body.encode('utf-8'))

microservice = HTTPServer((HOST_NAME, LISTEN_PORT), httpd)

while keep_running():
    microservice.handle_request()


microsimclient/microsimclient.py
#!/usr/local/bin/python3
import sys
import os
import socket
import time
import random
import string
import requests
import json
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer
from statsd import StatsClient

def str2bool(val):
    if val and val.lower() != 'false':
        return bool(val)
    return False

REQUEST_URLS = os.getenv('REQUEST_URLS', None)
REQUEST_INTERNET = str2bool(os.getenv('REQUEST_INTERNET', False))
REQUEST_MALWARE = str2bool(os.getenv('REQUEST_MALWARE', False))
SEND_SQLI = str2bool(os.getenv('SEND_SQLI', False))
SEND_DIR_TRAVERSAL = str2bool(os.getenv('SEND_DIR_TRAVERSAL', False))
SEND_XSS = str2bool(os.getenv('SEND_XSS', False))
SEND_DGA = str2bool(os.getenv('SEND_DGA', False))
REQUEST_WAIT_SECONDS = float(os.getenv('REQUEST_WAIT_SECONDS', 1.0))
REQUEST_BYTES = int(os.getenv('REQUEST_BYTES', 1024))
STOP_SECONDS = int(os.getenv('STOP_SECONDS', 0))
STOP_PADDING = str2bool(os.getenv('STOP_PADDING', False))
REQUEST_PROBABILITY = float(os.getenv('REQUEST_PROBABILITY', 1.0))
ATTACK_PROBABILITY = float(os.getenv('ATTACK_PROBABILITY', 0.01))
EGRESS_PROBABILITY = float(os.getenv('EGRESS_PROBABILITY', 0.1))
STATS_PORT = os.getenv('STATS_PORT', None)
STATSD_HOST = os.getenv('STATSD_HOST', None)
STATSD_PORT = int(os.getenv('STATSD_PORT', 8125))
START_TIME = int(time.time())
HOST_NAME = ''

url_list = REQUEST_URLS.split(',')

padding = 0
if STOP_SECONDS and STOP_PADDING:
    padding = random.choice(range(STOP_SECONDS))

attacks_selected = []
if SEND_DGA:
    attacks_selected.append('dga')
if SEND_SQLI:
    attacks_selected.append('sqli')
if SEND_DIR_TRAVERSAL:
    attacks_selected.append('dt')
if SEND_XSS:
    attacks_selected.append('xss')
if REQUEST_MALWARE:
    attacks_selected.append('malware')

egress_sites = [
    'http://mirror.facebook.net/centos/',
    'http://mirror.rackspace.com/CentOS/',
    'http://mirrors.edge.kernel.org/centos/',
    'http://mirrors.oit.uci.edu/centos/',
    'http://www.gtlib.gatech.edu/pub/centos/',
    'http://mirror.grid.uchicago.edu/pub/linux/centos/',
    'http://mirror.cc.columbia.edu/pub/linux/centos/',
    'http://yum.tamu.edu/centos/',
    'http://mirror.cogentco.com/pub/linux/centos/',
    'http://mirror.cs.vt.edu/pub/CentOS/',
    'http://centos.s.uw.edu/centos/'
]

dga_domains = [
    't3622c4773260c097e2e9b26705212ab85.ws',
    'u83ccf36d9f02e9ea79a9d16c0336677e4.to',
    'v02bec0c090508bc76b3ea81dfc2198a71.in',
    'wa9e4628c334324e181e40f33f878c153f.hk',
    'xdcc5481252db5f38d5fc18c9ad3b2f7fd.cn',
    'yf32d9ac7f0a9f463e8da4736b12d7044a.tk'
]

stats = {
    'Total': {
        'Requests': 0,
        'Sent Bytes': 0,
        'Received Bytes': 0,
        'Internet Requests': 0,
        'Attacks': 0,
        'SQLi': 0,
        'XSS': 0,
        'Directory Traversal': 0,
        'DGA': 0,
        'Malware': 0,
        'Error': 0
    },
    'Last 30 Seconds': {
        'Requests': 0,
        'Sent Bytes': 0,
        'Received Bytes': 0,
        'Internet Requests': 0,
        'Attacks': 0,
        'SQLi': 0,
        'XSS': 0,
        'Directory Traversal': 0,
        'DGA': 0,
        'Malware': 0,
        'Error': 0
    }
}

if STATSD_HOST:
    client_stats = StatsClient(prefix='all_clients',
                               host=STATSD_HOST,
                               port=STATSD_PORT)

    host_stats = StatsClient(prefix='client-' + socket.gethostname(),
                             host=STATSD_HOST,
                             port=STATSD_PORT)

lock = threading.Lock()

class httpd(BaseHTTPRequestHandler):
    server_name = socket.gethostname()
    server_ip = socket.gethostbyname(server_name)

    def do_GET(self):
        """json api response"""
        host_header = self.headers['Host']
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        with lock:
            self.response = {
                'time': time.asctime(),
                'runtime': int(time.time() - START_TIME),
                'hostname': self.server_name,
                'ip': self.server_ip,
                'stats': stats['Total'],
                'config': {
                    'STATS_PORT': int(STATS_PORT),
                    'STATSD_HOST': STATSD_HOST,
                    'STATSD_PORT': STATSD_PORT,
                    'REQUEST_URLS': REQUEST_URLS,
                    'REQUEST_INTERNET': REQUEST_INTERNET,
                    'REQUEST_MALWARE': REQUEST_MALWARE,
                    'SEND_SQLI': SEND_SQLI,
                    'SEND_DIR_TRAVERSAL': SEND_DIR_TRAVERSAL,
                    'SEND_XSS': SEND_XSS,
                    'SEND_DGA': SEND_DGA,
                    'REQUEST_WAIT_SECONDS': REQUEST_WAIT_SECONDS,
                    'REQUEST_BYTES': REQUEST_BYTES,
                    'STOP_SECONDS': STOP_SECONDS,
                    'STOP_PADDING': STOP_PADDING,
                    'TOTAL_STOP_SECONDS': STOP_SECONDS + padding,
                    'REQUEST_PROBABILITY': REQUEST_PROBABILITY,
                    'EGRESS_PROBABILITY': EGRESS_PROBABILITY,
                    'ATTACK_PROBABILITY': ATTACK_PROBABILITY
                }
            }
        body = json.dumps(self.response, indent=2)
        self.wfile.write(body.encode('utf-8'))

class state():
    last_timestamp = START_TIME

def every_30_seconds():
    with lock:
        if state.last_timestamp + 30 > int(time.time()):
            return False

        state.last_timestamp = int(time.time())
        return True

def keep_running():
    if not REQUEST_URLS:
        sys.exit('Server killed - REQUEST_URLS environment variable required.')

    if (STOP_SECONDS != 0) and ((START_TIME + STOP_SECONDS + padding) < int(time.time())):
        sys.exit('Server killed after ' + str(int(STOP_SECONDS) + int(padding)) + ' seconds.')

    return True

def insert_data():
    return ''.join([random.choice(string.ascii_letters + string.digits) for n in range(REQUEST_BYTES)])

def statistics_server():
    stats_server = HTTPServer((HOST_NAME, int(STATS_PORT)), httpd)
    stats_server.serve_forever()

def main():
    while keep_running():
        if every_30_seconds():
            # Print and clear statistics
            with lock:
                print(json.dumps(stats))
                stats['Last 30 Seconds'] = {
                    'Requests': 0,
                    'Sent Bytes': 0,
                    'Received Bytes': 0,
                    'Internet Requests': 0,
                    'Attacks': 0,
                    'SQLi': 0,
                    'XSS': 0,
                    'Directory Traversal': 0,
                    'DGA': 0,
                    'Malware': 0,
                    'Error': 0
                }

        json_body = {'data': insert_data()}

        def internal_request(url):
            try:
                response = requests.post(url, json=json_body, timeout=3)
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   Request to ' + response.url + '   Request size: ' + str(len(response.request.body)) + '   Response size: ' + str(len(response.content)))
                with lock:
                    stats['Total']['Requests'] += 1
                    stats['Last 30 Seconds']['Requests'] += 1
                    stats['Total']['Sent Bytes'] += len(response.request.body)
                    stats['Last 30 Seconds']['Sent Bytes'] += len(response.request.body)
                    stats['Total']['Received Bytes'] += len(response.content)
                    stats['Last 30 Seconds']['Received Bytes'] += len(response.content)

                if STATSD_HOST:
                    client_stats.incr('requests')
                    client_stats.incr('sent_bytes', len(response.request.body))
                    client_stats.incr('received_bytes', len(response.content))
                    host_stats.incr('requests')
                    host_stats.incr('sent_bytes', len(response.request.body))
                    host_stats.incr('received_bytes', len(response.content))

            except Exception as e:
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   ' + str(e) + ' error to: ' + url)
                with lock:
                    stats['Total']['Error'] += 1
                    stats['Last 30 Seconds']['Error'] += 1

                if STATSD_HOST:
                    client_stats.incr('error')
                    host_stats.incr('error')

        def sqli_attack(sqli_victim):
            parameters = {
                'username': 'joe@example.com',
                'password': ';UNION SELECT 1, version() limit 1,1--'
            }

            try:
                sqli = requests.get(sqli_victim, params=parameters, timeout=3)
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   SQLi sent: ' + sqli.url)
                with lock:
                    stats['Total']['SQLi'] += 1
                    stats['Last 30 Seconds']['SQLi'] += 1
                    stats['Total']['Attacks'] += 1
                    stats['Last 30 Seconds']['Attacks'] += 1

                if STATSD_HOST:
                    client_stats.incr('sqli')
                    client_stats.incr('attacks')
                    host_stats.incr('sqli')
                    host_stats.incr('attacks')

            except Exception as e:
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   ' + str(e) + ' error to: ' + sqli_victim)
                with lock:
                    stats['Total']['Error'] += 1
                    stats['Last 30 Seconds']['Error'] += 1

                if STATSD_HOST:
                    client_stats.incr('error')
                    host_stats.incr('error')

        def xss_attack(xss_victim):
            parameters = {
                'username': 'joe@example.com',
                'password': "pwd<script>alert('attacked')</script>"
            }

            try:
                xss = requests.get(xss_victim, params=parameters, timeout=3)
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   XSS sent: ' + xss.url)
                with lock:
                    stats['Total']['XSS'] += 1
                    stats['Last 30 Seconds']['XSS'] += 1
                    stats['Total']['Attacks'] += 1
                    stats['Last 30 Seconds']['Attacks'] += 1

                if STATSD_HOST:
                    client_stats.incr('xss')
                    client_stats.incr('attacks')
                    host_stats.incr('xss')
                    host_stats.incr('attacks')

            except Exception as e:
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   ' + str(e) + ' error to: ' + xss_victim)
                with lock:
                    stats['Total']['Error'] += 1
                    stats['Last 30 Seconds']['Error'] += 1

                if STATSD_HOST:
                    client_stats.incr('error')
                    host_stats.incr('error')

        def dt_attack(dt_victim):
            parameters = {
                'username': 'joe@example.com',
                'password': '../../../../../passwd'
            }

            try:
                dirtraversal = requests.get(dt_victim, params=parameters, timeout=3)
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   Directory Traversal sent: ' + dirtraversal.url)
                with lock:
                    stats['Total']['Directory Traversal'] += 1
                    stats['Last 30 Seconds']['Directory Traversal'] += 1
                    stats['Total']['Attacks'] += 1
                    stats['Last 30 Seconds']['Attacks'] += 1

                if STATSD_HOST:
                    client_stats.incr('directory_traversal')
                    client_stats.incr('attacks')
                    host_stats.incr('directory_traversal')
                    host_stats.incr('attacks')

            except Exception as e:
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   ' + str(e) + ' error to: ' + dt_victim)
                with lock:
                    stats['Total']['Error'] += 1
                    stats['Last 30 Seconds']['Error'] += 1

                if STATSD_HOST:
                    client_stats.incr('error')
                    host_stats.incr('error')

        def egress_request(egress_site):
            egress_internet = requests.Session()
                
            try:
                egress_internet = egress_internet.get(egress_site, allow_redirects=True, timeout=3)
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   Internet request to: ' + egress_internet.url)
                with lock:
                    stats['Total']['Internet Requests'] += 1
                    stats['Last 30 Seconds']['Internet Requests'] += 1
                    stats['Total']['Received Bytes'] += len(egress_internet.content)
                    stats['Last 30 Seconds']['Received Bytes'] += len(egress_internet.content)

                if STATSD_HOST:
                    client_stats.incr('internet_requests')
                    client_stats.incr('received_bytes', len(egress_internet.content))
                    host_stats.incr('internet_requests')
                    host_stats.incr('received_bytes', len(egress_internet.content))

            except Exception as e:
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   ' + str(e) + ' error to: ' + egress_site)
                with lock:
                    stats['Total']['Error'] += 1
                    stats['Last 30 Seconds']['Error'] += 1

                if STATSD_HOST:
                    client_stats.incr('error')
                    host_stats.incr('error')
                
        def malware_request():
            eicar = requests.Session()
            eicar_url = 'http://www.eicar.org/download/eicar.com.txt'

            try:
                eicar = eicar.get(eicar_url, timeout=3)
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   Malware downloaded: ' + eicar.text)
                with lock:
                    stats['Total']['Malware'] += 1
                    stats['Last 30 Seconds']['Malware'] += 1
                    stats['Total']['Attacks'] += 1
                    stats['Last 30 Seconds']['Attacks'] += 1

                if STATSD_HOST:
                    client_stats.incr('malware')
                    client_stats.incr('attacks')
                    host_stats.incr('malware')
                    host_stats.incr('attacks')

            except Exception as e:
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   ' + str(e) + ' error to: ' + eicar_url)
                with lock:
                    stats['Total']['Error'] += 1
                    stats['Last 30 Seconds']['Error'] += 1

                if STATSD_HOST:
                    client_stats.incr('error')
                    host_stats.incr('error')

        def dga_attack(dga):
            try:
                dga_response = socket.gethostbyname(dga)
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   DGA query sent: ' + dga + '   Response: ' + dga_response)
                with lock:
                    stats['Total']['DGA'] += 1
                    stats['Last 30 Seconds']['DGA'] += 1
                    stats['Total']['Attacks'] += 1
                    stats['Last 30 Seconds']['Attacks'] += 1
                
                if STATSD_HOST:
                    client_stats.incr('dga')
                    client_stats.incr('attacks')
                    host_stats.incr('dga')
                    host_stats.incr('attacks')

            except Exception as e:
                print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   ' + str(e) + ' error resolving: ' + dga)
                with lock:
                    stats['Total']['Error'] += 1
                    stats['Last 30 Seconds']['Error'] += 1

                if STATSD_HOST:
                    client_stats.incr('error')
                    host_stats.incr('error')

        # Put each internal request in its own thread
        if random.random() < REQUEST_PROBABILITY:
            url = random.choice(url_list)
            request_thread = threading.Thread(target=internal_request, args=(url,), daemon=True)
            request_thread.start()
        else:
            print(time.strftime("%Y-%m-%dT%H:%M:%S") + '   Skipped internal request this run.')

        # Put each external internet request in its own thread
        if REQUEST_INTERNET:
            if random.random() < EGRESS_PROBABILITY:
                egress_site = random.choice(egress_sites)
                egress_thread = threading.Thread(target=egress_request, args=(egress_site,), daemon=True)
                egress_thread.start()

        # Select only one attack per loop
        if attacks_selected:
            do_attack = random.choice(attacks_selected)

            # Put each attack in its own thread
            if random.random() < ATTACK_PROBABILITY:
                if do_attack == 'dga':
                    dga = random.choice(dga_domains)
                    dga_thread = threading.Thread(target=dga_attack, args=(dga,), daemon=True)
                    dga_thread.start()

                if do_attack == 'sqli':
                    sqli_victim = random.choice(url_list)
                    sqli_thread = threading.Thread(target=sqli_attack, args=(sqli_victim,), daemon=True)
                    sqli_thread.start()

                if do_attack == 'malware':
                    malware_thread = threading.Thread(target=malware_request, daemon=True)
                    malware_thread.start()

                if do_attack == 'dt':
                    dt_victim = random.choice(url_list)
                    dt_thread = threading.Thread(target=dt_attack, args=(dt_victim,), daemon=True)
                    dt_thread.start()

                if do_attack == 'xss':
                    xss_victim = random.choice(url_list)
                    xss_thread = threading.Thread(target=xss_attack, args=(xss_victim,), daemon=True)
                    xss_thread.start()

        time.sleep(REQUEST_WAIT_SECONDS)

if STATS_PORT:
    stats_thread = threading.Thread(target=statistics_server, daemon=True)
    stats_thread.start()

main()
